hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -433,7 +442,7 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
   }
 
   private def constructOrCopy(mb: MethodBuilder, region: Code[Region], srcArray: PArray, srcAddress: Code[Long], forceDeep: Boolean): Code[Long] = {
-    if (srcArray == this) {
+    if (srcArray.elementType == elementType) {
```

Comment:
I see why you made this change (avoid unnecessary copies) but it makes me nervous -- when we add another implementation of PArray, this will break.

Reply:
We should add a method `sameExceptTopLevelRequiredness` or something, which is what this is trying to ask, yes?

Reply:
Yes, broken.  I added an `isInstanceOf[PCanonicalArray]`.